scipy/linalg/_solvers.py

```diff
@@ -80,6 +80,18 @@ def solve_sylvester(a, b, q):
     True
 
     """
+    for ind, _ in enumerate((a, b)):
+        if not np.equal(*_.shape):
+            raise ValueError("Matrix {} should be square.".format("ab"[ind]))
+    # Shape consistency check
+    if a.shape[0] != q.shape[0] or b.shape[0] != q.shape[1]:
+        raise ValueError("Matrix q should have shape (a.shape[0], b.shape[0])")
```

Comment:
Generic messages in my opinion serve no help. The error message here is clearly a bug but it should tell what the issue is. `(5, 4)` here is exactly what user should check in their own problem but the message is missing why it should be `(5,4)` otherwise `wrong size` means nothing and does not answer "What is wrong with it?" question.

Reply:
I didn't write what the message was going to be. Perhaps you would have found it to be helpful. But instead, I'll go ahead and de-scope this PR to just address gh-20372. It was a nice thought by the author to also improve the input validation in the same PR, but is orthogonal to the main focus. Since it would essentially start from scratch anyway - rewrite this, add tests, etc... - I will go ahead and remove it so we can get this merged.

Reply:
I'm just saying if we are going to touch it, we can make it more informative; not trying to start a bikeshed. But ok I won't clutter.